### PR TITLE
fix #1779: show critical path latency for application 

### LINKED ISSUE #1779: overall Average Message Receive Latency should not be averaged over Source Processors
since Average Message Receive Latency of a Source Processor is always 0

### PR DESCRIPTION
**Pull request by stanleyxu2005**
On application overview page, it will show 3 charts (send message throughput, receive message throughput and critical path latency). On processor detail page, it will continue show 4 charts as before.

Second the PR has overhauled the model package. More cleaner code separation

And the PR contains some ui changes:

1. Chart's tooltip will disappear when mouse is out of the chart control area
2. Empty table (e.g. empty application list table) now has a better description panel
3. Click on table column's sort arrow will have animation effect
4. Pagination control would not update when the table is showing some search results.
5. The first hour application will not receive any historical metrics, so the chart for "Past 72 Hours" will show "No Graph Data Available".

### CODE REVIEW

File: services/dashboard/views/apps/streamingapp/streamingapp.js [hunk 1]
```diff
@@ -101,9 +99,9 @@ angular.module('dashboard')
       $scope.sendThroughputMetricsDescription = 'Messages are sent from external data source';
       $scope.receiveThroughputMetricsCaption = 'Sink Processors Receive Throughput';
       $scope.receiveThroughputMetricsDescription = 'Messages are received by external data sink';
+      $scope.messageLatencyMetricsCaption = 'Critical Path Latency';
+      $scope.messageLatencyMetricsDescription = 'The total latency on the critical path';
```
Comment (clockfly) [inline services/dashboard/views/apps/streamingapp/streamingapp.js]: Should this description be more clear?

File: services/dashboard/services/models/dag.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      this.degrees = this._calculateDegrees();
+    }
+
+    Dag.prototype = {
+
+      isTopologyChanged: function(processors, edges) {
+        return !_.isEqual(this.processors, processors) || !_.isEqual(this.edges, edges);
+      },
+
+      getProcessorIds: function() {
```
Comment (clockfly) [inline services/dashboard/services/models/dag.js]: As this is a util, can you document the return type.

File: services/dashboard/services/models/dag.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        }, this);
+      },
+
+      getSinkProcessorIds: function() {
+        return _.filter(this.getProcessorIds(), function(processorId) {
+          return this.degrees[processorId].outdegree === 0;
+        }, this);
+      },
+
+      hierarchyDepth: function() {
```
Comment (clockfly) [inline services/dashboard/services/models/dag.js]: As this is a util, can you document the return type.

File: services/dashboard/services/models/dag.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+        _.forEach(this.edges, function(edge) {
+          result[edge.from].outdegree++;
+          result[edge.to].indegree++;
+        });
+        return result;
+      },
+
+      /** Return the critical path. */
+      calculateCriticalPath: function(metricsProvider, time) {
```
Comment (clockfly) [inline services/dashboard/services/models/dag.js]: As this is a util, can you document the return type.

File: services/dashboard/views/apps/streamingapp/streamingapp.js (hunk 1 above)
Comment (clockfly) [inline services/dashboard/views/apps/streamingapp/streamingapp.js]: Suggest to change "Critical Path Latency" to more user friendly name "end to end latency", and in the notes document how end to end latency is caluculated (by critical path)

File: services/dashboard/services/models/dag.js (hunk 4 above)
Comment (stanleyxu2005) [inline services/dashboard/services/models/dag.js]: Yes. The javascript does not hint return type, if you are not using a *good* ide. So I'll describe return type of public methods, those method name does not imply its return type obviously.

**Comment by stanleyxu2005** [comment]
@clockfly Added more docstring and changed the description of critical path latency. Please kindly review.

**Comment by clockfly** [comment]
+1 for most parts. 
Few suggestion:
1. handle negative latency value (suppose network not synced) more neatly. 
2. Better naming of function name.
3. Clearer tooltip for soure send throughput and sink receive throughput.
4. Try to name the anonymous function in forEach.

**Pull request merged**